elastic/kibana
# [Security Solution] Rule Diff Phase 2 components

## Summary

Addresses https://github.com/elastic/kibana/issues/166489
Docs issue: https://github.com/elastic/security-docs/issues/4783

Adds per-field diffs for the rule upgrade flyout 

### Acceptance Criteria

- [x] The tab with per-field diffs is hidden behind a new feature flag. When the flag is off, the tab does not appear in the flyout. The tab should work regardless of the value of `jsonPrebuiltRulesDiffingEnabled`.
- [x] Per-field diffs are read-only components. We don't need to let the user "merge" differences using these components.
- [x] Diffs for complex fields are rendered as JSON diffs using the same component used for rendering the JSON diff for the whole rule. This means this component should be abstracted away and should accept `unknown` values in props instead of `RuleResponse`.
- [x] Diffs for related fields are grouped or rendered close to each other. For example:
  - [x] Index patterns + Data view id
  - [x] Custom query + Filters + Language + Saved query id
- [x] The tab uses the response from the `upgrade/_review` API endpoint and doesn't need any other API calls to render itself.
- [x] The tab renders itself under 150ms.

### Manual testing

We are currently going over every field returned by the rule diff api to locate any issues (either in the UI or endpoint) that occur when certain fields are **added**/**deleted**/**modified** 

To test certain fields: 

Use either a `PATCH` or `PUT` method on the `/api/detection_engine/rules` endpoint route with a rule object that contains at least a `rule_id` (not `id`) and a lower version number than the current rule's version. This will allow you to see the prebuilt package update in the Rule Updates tab on the rules overview page

Example
```json
{
  "rule_id": "some_rule_id_12345",
  "version": 1,
  "test_field": "test value",
  ...
}
```

<details>

  <summary>Manually tested fields</summary>

| Field name                | Has been verified  | Endpoint issue |
|---------------------------|--------------------|----------------|
| `version`                 | :heavy_check_mark: |                |
| `name`                    | :heavy_check_mark: |             |
| `description`             | :heavy_check_mark: |            |
| `author`                  | :heavy_check_mark: |            |
| `building_block`          | :heavy_check_mark: |                |
| `severity`                | :heavy_check_mark: |                |
| `severity_mapping`        | :heavy_check_mark: |                |
| `risk_score`              |:heavy_check_mark:|                |
| `risk_score_mapping`      | :heavy_check_mark: |                |
| `references`              | :heavy_check_mark: |                |
| `false_positives`         | :heavy_check_mark: |                |
| `investigation_fields`    | :heavy_check_mark:  |                |
| `license`                 | :heavy_check_mark: |            |
| `rule_name_override`      | :heavy_check_mark: |                |
| `threat`                  |:heavy_check_mark: |           |
| `threat_indicator_path`   | :heavy_check_mark: |                |
| `timestamp_override`      | :heavy_check_mark: |                |
| `tags`                    |:heavy_check_mark:|                |
| `data_source`             |:heavy_check_mark:|                |
| `type`                    |:heavy_check_mark:|                |
| `kql_query`               |:heavy_check_mark:|                |
| `eql_query`               |:heavy_check_mark:|                |
| `event_category_override` | :heavy_check_mark: |                |
| `timestamp_field`         | :heavy_check_mark:  |                |
| `tiebreaker_field`        | :heavy_check_mark:  |                |
| `esql_query`              | :heavy_check_mark: |                |
| `anomaly_threshold`       |:heavy_check_mark:|                |
| `machine_learning_job_id` | :heavy_check_mark: |                |
| `related_integrations`    | :heavy_check_mark: |                |
| `required_fields`         | :heavy_check_mark: |                |
| `timeline_template`       | :heavy_check_mark: |                |
| `threshold`               |:heavy_check_mark:|                |
| `threat_index`            | :heavy_check_mark: |                |
| `threat_mapping`          | :heavy_check_mark: |                |
| `threat_filters`          | :heavy_check_mark:  |                |
| `threat_query`            |  :heavy_check_mark: |                |
| `threat_indicator_path`   | :heavy_check_mark: |                |
| `concurrent_searches`     |  :heavy_check_mark: |                |
| `items_per_search`        | :heavy_check_mark: |                |
| `alert_suppression`       | :heavy_check_mark: |                |
| `new_terms_fields`        | :heavy_check_mark: |                |
| `history_window_start`    | :heavy_check_mark: |                |
| `rule_schedule`           | :heavy_check_mark: |                |
| `setup`                   | :heavy_check_mark: |                |
| `note`                    |:heavy_check_mark:|                |
| `max_signals`             | :heavy_check_mark: |                |



</details>

### Screenshots

<img width="1587" alt="Screenshot 2024-02-07 at 1 36 34 AM" src="https://github.com/elastic/kibana/assets/56367316/85dce529-064e-4025-b82c-2e89f6ec800b">
<img width="994" alt="Screenshot 2024-02-07 at 1 36 52 AM" src="https://github.com/elastic/kibana/assets/56367316/c226973f-ad46-4565-90c0-437316b138b4">

### Checklist

Delete any items that are not applicable to this PR.



### For maintainers
